OpenDreamRuntime/Procs/DebugAdapter/DreamDebugManager.cs

```diff
@@ -105,10 +106,56 @@ private struct ActiveBreakpoint {
         Output(message, category);
     }
 
+    public void HandleInstruction(DMProcState state) {
+        // Handle proc starts and instruction stepping as one, so instruction
+        // stepping after a function breakpoint doesn't stop at the
+        // instruction again.
+
+        // If the PC is 0, check for a function breakpoint.
+        if (state.ProgramCounter == 0) {
+            var hit = new List<int>();
+            if (_possibleFunctionBreakpoints?.GetValueOrDefault((state.Proc.OwningType.PathString, state.Proc.Name)) is FunctionBreakpointSlot slot) {
+                foreach (var bp in slot.Breakpoints) {
+                    if (TestBreakpoint(bp)) {
+                        hit.Add(bp.Id);
+                    }
+                }
+            }
+            if (hit.Any()) {
+                Output($"Function breakpoint hit at {state.Proc.OwningType.PathString}::{state.Proc.Name}");
+                Stop(state.Thread, new StoppedEvent {
+                    Reason = StoppedEvent.ReasonFunctionBreakpoint,
+                    HitBreakpointIds = hit,
+                });
+                return;
+            }
+        }
+
+        // Stop if we're instruction stepping.
+        bool stoppedOnStep = false;
+        switch (_threadStepModes.GetValueOrDefault(state.Thread.Id)) {
```

Comment:
This extra dictionary index on every instruction is slowing things down. Is there some other way to do this?

Reply:
Or worse case scenario only do it in debug?

Reply:
- It could be stored on DreamThread directly, if you're willing for debugger state to leak there.
- We could skip the check if no debug client is attached, or if _threadStepModes is empty, or similar.
- I would not recommend putting it behind `#if DEBUG` since some may want to debug their DM code without debugging OD itself.

Reply:
I'm happy with that DreamThread idea.